osquery/events/linux/http_event_publisher.cpp

```diff
@@ -0,0 +1,863 @@
+ #define OFFSET_CIPHER_LIST      44
+  
+ /* Grease bytes to ignore */
+    std::set<std::string> greaseBytes = {"0A0A","1A1A","2A2A","3A3A","4A4A","5A5A","6A6A","7A7A",
+                                         "8A8A","9A9A","AAAA","BABA","CACA","DADA","EAEA","FAFA"};
+    /* Implementation of whitelisting  */
+    /*
+     * @brief whitelisting of headers - to be included in other_headers column
+     *
+     * By default no headers are inlcuded in headers list.
+     * headers specify them via this flag as a comma separated list.
+     */
+ FLAG(string,
+      include_http_headers,
+      "",
+      "Comma-separated list of headers to be included in http_events table");
+ 
+ /* fingerprint struct */
+ struct fingerprint {
+      uint16_t  extensions_length;
+      uint8_t   *extensions;
+      uint16_t  curves_length;
+      uint8_t   *curves;
+      uint16_t  sig_alg_length;
+      uint8_t   *sig_alg;
+      uint16_t  ec_point_fmt_length;
+      uint8_t   *ec_point_fmt;
+ };
+    
+ const char* ssl_version(u_short version) {
+     static char hex[7];
+     switch (version) {
+         case 0x002: return "SSLv2";
+         case 0x300: return "SSLv3";
+         case 0x301: return "TLSv1.0";
+         case 0x302: return "TLSv1.1";
+         case 0x303: return "TLSv1.2";
+         case 0x304: return "TLSv1.3";
+     }
+     snprintf(hex, sizeof(hex), "0x%04hx", version);
+     return hex;
+ }
+    
+
+ REGISTER(HTTPLookupEventPublisher, "event_publisher", "http_lookups");
+ 
+ FLAG(bool,
+      enable_http_lookups,
+      false,
+      "Enable the HTTP capture event publisher");
+
+ FLAG(string,
+      disable_http_event_filters,
+      "",
+      "Comma-separated list of filters(e.g. "
+      "'private_ip_events') to disable specific default http filter.");
+
+ Status HTTPLookupEventPublisher::setUp() {
+   if (!FLAGS_enable_http_lookups) {
+     return Status(1, "HTTP lookups publisher disabled via configuration.");
+   }
+
+   // PCAP has no payload[offset] field, so we need to get the payload offset
+   // from the TCP header (offset 12, upper 4 bits, number of 4-byte words):
+   // TLS Handshake starts with a '22' byte, version, length,
+   // and then '01'/'02' for client/server hello
+   // TLS Handshake starts with a '22' byte, version, length,
+   // and then '01'/'02' for client/server hello
+
+   httpFilter_.clear();
+   httpFilter_ = {{"http_request", kHttpRequestFilter + kTLSTrafficFilter},
+                  {"private_ip_events", kInternalTrafficFilter}};
+   return Status(0, "OK");
+ }
+
+ void HTTPLookupEventPublisher::getFilters(std::string& sFilters) {
+   for (auto& filter_name : osquery::split(FLAGS_disable_http_event_filters, ",")) {
+     // To enforce unification we lower case all filters for matching
+     std::transform(filter_name.begin(),
+                    filter_name.end(),
+                    filter_name.begin(),
+                    ::tolower);
+
+     size_t pos = 0;
+     while (pos < httpFilter_.size()) {
```

Comment:
This while loop can be replaced with a find() + erase(). It can be further simplified by making httpFilter_ read only (pseudo code)

```cpp
Status HTTPLookupEventPublisher::getFilters(std::string& filter) {
  filter.clear();

  std::unordered_set<std::string> disabled_filter_list;

  for (auto disabled_filter : osquery::split(FLAGS_disable_http_event_filters, ",")) {
    std::transform(disabled_filter, ...)
    disabled_filter_list.insert(std::move(disabled_filter));
  }

  std::string output;

  for (const auto &filter_pair : kDefaultFilterList) { // httpFilter_
    auto filter_name = filter_pair.first;
    std::transform(filter_name, ...);

    if (disabled_filter_list.count(filter_name) > 0U) {
      continue;
    }

    const auto &filter_logic.= filter_pair.second;
    filter += filter_logic ........
  }

  if (output.empty()) {
    return Status::failure(".....");
  }

  filter = std::move(output);
  return Status::success();
}
```

Functions should always return Status if there is a chance they may fail. Variable names should follow snake_case in new code

Reply:
Removed this function itself as I am monitoring internal IP traffic as well now, based on @directionless comment and review.